cluster-controller/src/main/java/io/strimzi/controller/cluster/resources/TopicController.java

```diff
@@ -0,0 +1,245 @@
+    protected static final int DEFAULT_BOOTSTRAP_SERVERS_PORT = 9092;
+    protected static final String DEFAULT_FULL_RECONCILIATION_INTERVAL = "15 minutes";
+    protected static final String DEFAULT_ZOOKEEPER_SESSION_TIMEOUT = "20 seconds";
+
+    // Configuration keys
+    public static final String KEY_CONFIG = "topic-controller-config";
+
+    // Topic Controller configuration keys
+    public static final String KEY_KAFKA_BOOTSTRAP_SERVERS = "STRIMZI_KAFKA_BOOTSTRAP_SERVERS";
+    public static final String KEY_ZOOKEEPER_CONNECT = "STRIMZI_ZOOKEEPER_CONNECT";
+    public static final String KEY_NAMESPACE = "STRIMZI_NAMESPACE";
+    public static final String KEY_FULL_RECONCILIATION_INTERVAL = "STRIMZI_FULL_RECONCILIATION_INTERVAL";
+    public static final String KEY_ZOOKEEPER_SESSION_TIMEOUT = "STRIMZI_ZOOKEEPER_SESSION_TIMEOUT";
+
+    /**
+     * Constructor
+     *
+     * @param namespace Kubernetes/OpenShift namespace where cluster resources are going to be created
+     * @param cluster   overall cluster name
+     */
+    protected TopicController(String namespace, String cluster) {
+
+        super(namespace, cluster);
+        this.name = topicControllerName(cluster);
+        this.image = DEFAULT_IMAGE;
+        this.replicas = DEFAULT_REPLICAS;
+        this.healthCheckPath = "/";
+        this.healthCheckTimeout = DEFAULT_HEALTHCHECK_TIMEOUT;
+        this.healthCheckInitialDelay = DEFAULT_HEALTHCHECK_DELAY;
+
+        // create a default configuration
+        this.config = new TopicControllerConfig();
+        this.config
+                .withImage(DEFAULT_IMAGE)
+                .withKafkaBootstrapServers(defaultBootstrapServers(cluster))
+                .withZookeeperConnect(defaultZookeeperConnect(cluster))
+                .withNamespace(namespace)
+                .withReconciliationInterval(DEFAULT_FULL_RECONCILIATION_INTERVAL)
+                .withZookeeperSessionTimeout(DEFAULT_ZOOKEEPER_SESSION_TIMEOUT);
+    }
+
+    public static String topicControllerName(String cluster) {
+        return cluster + TopicController.NAME_SUFFIX;
+    }
+
+    protected static String defaultZookeeperConnect(String cluster) {
+        return cluster + ZookeeperCluster.NAME_SUFFIX + ":" + DEFAULT_ZOOKEEPER_PORT;
+    }
+
+    protected static String defaultBootstrapServers(String cluster) {
+        return cluster + KafkaCluster.NAME_SUFFIX + ":" + DEFAULT_BOOTSTRAP_SERVERS_PORT;
+    }
+
+    public void setConfig(TopicControllerConfig config) {
+        this.config = config;
+    }
+
+    public TopicControllerConfig getConfig() {
+        return this.config;
+    }
+
+    /**
+     * Create a Topic Controller from the related ConfigMap resource
+     *
+     * @param kafkaClusterCm ConfigMap with cluster configuration containing the topic controller one
+     * @return Topic Controller instance, null if not configured in the ConfigMap
+     */
+    public static TopicController fromConfigMap(ConfigMap kafkaClusterCm) {
```

Comment:
Why do we use here `fromConfigMap` but `Storage` is using `fromJson`?

Reply:
I don't get your point sorry. The corresponding of `Storage` should be `TopicControllerConfig` which has `fromJson`.

Reply:
Seems a bit strange. Why do we have two classes here then?

Reply:
Sorry @scholzj ... are you speaking about why we have `TopicController` and `TopicControllerConfig` ?

Reply:
Yes. We don't have that anywhere else.

Reply:
So you are saying the the `TopicController` configuration should be inside the class itself not using the `TopicControllerConfig` class. I can agree with that because it's like the configuration we have for the `KafkaCluster` which is inside the class itself (but not the `Storage` which has a different class used inside the `KafkaCluster`). This point I didn't get right what you said on #119 : 

> Yeah, I think having the possibility to switch it on/off should be there. An additional configuration similar to storage might make sense (Docker image, timeouts, namespace? etc.).

And maybe we want to remove the `Storage` class as well and putting all its configuration in the `AbstractCluster` class directly ?

Reply:
I don't have problem with the `Storage` class. It is basically just an utility class integrated into `KafkaCluster` and `ZookeeperCluster`but not `KafkaConnectCluster`. I think it serves its purpose. But it is only a utility class used for a small part of the configuration which is shared. Here you created a new _top level_ class `TopicController` which is on the same level as `KafkaCluster` or `ZookeeperCluster`. That is fine. But for some reason you - unlike it is done in `KafkaCluster` or `ZookeeperCluster` - move the configuration into yet another class. I don't really have a big issue with it - it just seems inconsistent.

Reply:
I got your point now and I can agree on that but at this point my question is : to be "consistent" we should in some way : 

- removing `Storage` and having all its configuration just inside `AbstractCluster`

or

- thinking about more grouping for the other configuration (i.e. health checks, metrics, ...) making something like `Storage`.
In the second case we could svoid to have `AbstractCluster` exploding with more other fields over time.

Reply:
I don't think the `Storage` is a problem - although I agree that the metrics and healthcheck could be maybe done in the same way. But there is not `KafkaClusterConfig` class which would contain the things like Kafka configuration, Kafka docker image etc.

Reply:
Even without having `KafkaClusterConfig` for other things like Kafka config, docker image and so on, grouping the other could be better. I don't know if the configuration could explode in more fields in the `AbstractCluster` (and derived) in the future; I don't like class with a lot of fields that could be logically grouped instead. 
It's not something that we should change right now of course; I can remove the `TopicControllerConfig` and moving the config into the `TopicController` class but in any case I think that other configs could be handle as with `Storage` and don't think that it's a "mistake" as mentioned by @tombentley in another comment.